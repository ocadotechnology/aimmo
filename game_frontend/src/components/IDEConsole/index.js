import styled from 'styled-components'
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import LogEntries from 'components/LogEntries'

export const StyledConsole = styled.div`
  grid-area: ide-console;
  color: ${props => props.theme.palette.text.primary};
  background-color: ${props => props.theme.palette.background.default};
  font-family: ${props => props.theme.additionalVariables.typography.code.fontFamily};
  overflow: auto;
  white-space: pre-line;
  padding-bottom: 0px;

  ::-webkit-scrollbar {
    background-color: ${props => props.theme.palette.divider};
  }

  ::-webkit-scrollbar-track {
    background: ${props => props.theme.palette.divider}; 
  }

  ::-webkit-scrollbar-thumb {
    background: ${props => props.theme.palette.grey['A200']};
    border-radius: 100px;
    border: ${props => props.theme.spacing.unit / 4}px solid transparent;
    background-clip: content-box;
  }
`

export default class IDEConsole extends Component {
  state = {
    scrolledToBottom: true
  }

  handleScroll = event => {
    let srcElement = event.srcElement
    this.setState({...this.state, scrolledToBottom: srcElement.offsetHeight + srcElement.scrollTop === srcElement.scrollHeight})
  }

  componentDidMount () {
    if (this.consoleRef) {
      this.consoleRef.addEventListener('scroll', this.handleScroll)
    }
  }

  componentWillUnmount () {
    this.consoleRef.removeEventListener('scroll', this.handleScroll)
  }

  componentDidUpdate () {
    if (this.lastLogRef && this.state.scrolledToBottom) {
      this.lastLogRef.scrollIntoView(false)
    }
  }

  render () {
    return (
      <StyledConsole innerRef={ref => { this.consoleRef = ref }}>
        <LogEntries
          logs={this.props.logs}
          lastLogRef={ref => { this.lastLogRef = ref }} />
      </StyledConsole>
    )
  }
}

IDEConsole.propTypes = {
  logs: PropTypes.arrayOf(PropTypes.object)
}
