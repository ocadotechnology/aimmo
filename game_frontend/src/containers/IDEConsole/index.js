import styled from 'styled-components'
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import LogEntries from 'components/LogEntries'
import ConsoleBar from 'components/ConsoleBar'
import { connect } from 'react-redux'
import { actions } from 'redux/features/ConsoleLog'

export const IDEConsoleSection = styled.section`
  grid-area: ide-console;
  display: flex;
  flex-direction: column;
`

export const StyledConsole = styled.div`
  color: ${props => props.theme.palette.text.primary};
  background-color: ${props => props.theme.palette.background.default};
  font-family: ${props => props.theme.additionalVariables.typography.code.fontFamily};
  overflow: auto;
  white-space: pre-line;
  height: 100%;
  
  ::-webkit-scrollbar {
    background-color: ${props => props.theme.palette.divider};
  }

  ::-webkit-scrollbar-track {
    background: ${props => props.theme.palette.divider}; 
  }

  ::-webkit-scrollbar-thumb {
    background: ${props => props.theme.palette.grey['A200']};
    border-radius: 100px;
    border: ${props => props.theme.spacing(0.25)}px solid transparent;
    background-clip: content-box;
  }
`

export class IDEConsole extends Component {
  static propTypes = {
    logs: PropTypes.arrayOf(PropTypes.object)
  }

  state = {
    scrolledToBottom: true
  }

  handleScroll = event => {
    let target = event.target || event.srcElement
    this.setState({ ...this.state, scrolledToBottom: target.offsetHeight + target.scrollTop === target.scrollHeight })
  }

  componentDidMount () {
    if (this.consoleRef) {
      this.consoleRef.addEventListener('scroll', this.handleScroll)
    }
  }

  componentWillUnmount () {
    this.consoleRef.removeEventListener('scroll', this.handleScroll)
  }

  componentDidUpdate () {
    if (this.lastLogRef && this.state.scrolledToBottom) {
      this.lastLogRef.scrollIntoView(false)
    }
  }
  render () {
    return (
      <IDEConsoleSection>
        <ConsoleBar clearConsoleClicked={this.props.clearConsoleLogs} />
        <StyledConsole innerRef={ref => { this.consoleRef = ref }}>
          <LogEntries
            logs={this.props.logs}
            lastLogRef={ref => { this.lastLogRef = ref }} />
        </StyledConsole>
      </IDEConsoleSection>
    )
  }
}

const mapStateToProps = state => ({
  logs: state.consoleLog.logs
})

const mapDispatchToProps = {
  clearConsoleLogs: actions.clearConsoleLogs
}

export default connect(mapStateToProps, mapDispatchToProps)(IDEConsole)
