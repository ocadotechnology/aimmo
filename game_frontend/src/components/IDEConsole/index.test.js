/* eslint-env jest */
import React from 'react'
import IDEConsole from 'components/IDEConsole'
import createShallowWithTheme from 'testHelpers/createShallow'

describe('<IDEConsole />', () => {
  it('renders correctly', () => {
    const tree = createShallowWithTheme(<IDEConsole />)
    expect(tree).toMatchSnapshot()
  })
})
