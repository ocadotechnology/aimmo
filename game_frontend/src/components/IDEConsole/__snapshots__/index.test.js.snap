// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IDEConsole /> renders correctly 1`] = `
<styled.div
  innerRef={[Function]}
>
  <LogEntries
    lastLogRef={[Function]}
    logs={
      Array [
        Object {
          "log": "hello",
          "timestamp": "1",
        },
        Object {
          "log": "bye",
          "timestamp": "2",
        },
      ]
    }
  />
</styled.div>
`;

exports[`<StyledConsole /> renders correctly 1`] = `
.c0 {
  grid-area: ide-console;
  color: #fff;
  background-color: #303030;
  font-family: 'Source Code Pro',monospace;
  overflow: auto;
  white-space: pre-line;
  padding-bottom: 0px;
}

.c0::-webkit-scrollbar {
  background-color: rgba(255,255,255,0.12);
}

.c0::-webkit-scrollbar-track {
  background: rgba(255,255,255,0.12);
}

.c0::-webkit-scrollbar-thumb {
  background: #aaaaaa;
  border-radius: 100px;
  border: 2px solid transparent;
  background-clip: content-box;
}

<div
  className="c0"
/>
`;
